server/server.go

```diff
@@ -6,50 +6,185 @@ package server
-	reply := &messages.PublishReply{}
-	for _, evt := range req.Events {
-		serv.logger.Infof("Got event: %s", evt.String())
-		entryID, err := serv.queue.Publish(evt)
+// NewShipperServer creates a new server instance for handling gRPC endpoints.
+func NewShipperServer(publisher Publisher) (ShipperServer, error) {
+	if publisher == nil {
+		return nil, errors.New("publisher cannot be nil")
+	}
+
+	id, err := uuid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+
+	s := shipperServer{
+		uuid:      id.String(),
+		logger:    logp.NewLogger("shipper-server"),
+		publisher: publisher,
+		close:     &sync.Once{},
+	}
+
+	s.ctx, s.stop = context.WithCancel(context.Background())
+
+	return &s, nil
+}
+
+// GetAcceptedIndex returns the accepted index
+func (serv *shipperServer) GetAcceptedIndex() uint64 {
+	return uint64(serv.publisher.AcceptedIndex())
+}
+
+// GetPersistedIndex returns the persisted index
+func (serv *shipperServer) GetPersistedIndex() uint64 {
+	return uint64(serv.publisher.PersistedIndex())
+}
+
+// PublishEvents is the server implementation of the gRPC PublishEvents call.
+func (serv *shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
+	resp := &messages.PublishReply{
+		Uuid: serv.uuid,
+	}
+
+	// the value in the request is optional
+	if req.Uuid != "" && req.Uuid != serv.uuid {
+		resp.AcceptedIndex = serv.GetAcceptedIndex()
+		resp.PersistedIndex = serv.GetPersistedIndex()
+		serv.logger.Debugf("shipper UUID does not match, all events rejected. Expected = %s, actual = %s", serv.uuid, req.Uuid)
+
+		return resp, nil
+	}
+
+	for _, e := range req.Events {
+		_, err := serv.publisher.Publish(e)
+		if err == nil {
+			resp.AcceptedCount++
+			continue
+		}
+
+		if errors.Is(err, queue.ErrQueueIsFull) {
+			serv.logger.Debugf("queue is full, not all events accepted. Events = %d, accepted = %d", len(req.Events), resp.AcceptedCount)
+		} else {
+			err = fmt.Errorf("failed to enqueue an event. Events = %d, accepted = %d: %w", len(req.Events), resp.AcceptedCount, err)
+			serv.logger.Error(err)
+		}
+
+		break
+	}
+
+	resp.AcceptedIndex = serv.GetAcceptedIndex()
+	resp.PersistedIndex = serv.GetPersistedIndex()
+
+	serv.logger.
+		Debugf("finished publishing a batch. Events = %d, accepted = %d, accepted index = %d, persisted index = %d",
+			len(req.Events),
+			resp.AcceptedCount,
+			resp.AcceptedIndex,
+			resp.PersistedIndex,
+		)
+
+	return resp, nil
+}
+
+// PublishEvents is the server implementation of the gRPC PersistedIndex call.
+func (serv *shipperServer) PersistedIndex(req *messages.PersistedIndexRequest, producer pb.Producer_PersistedIndexServer) error {
+	serv.logger.Debug("new subscriber for persisted index change")
+	defer serv.logger.Debug("unsubscribed from persisted index change")
+
+	persistedIndex := serv.GetPersistedIndex()
+	if persistedIndex != 0 {
```

Comment:
Is skipping a 0 index just an optimization for the first iteration? It might be simpler to include it regardless, but probably we should at least send it anyway if `pollingIntervalDur == 0`.

Reply:
We send a response, please see above. As I stated in the documentation if the interval is zero we don't subscribe the client but rather send a single response and close the stream https://github.com/elastic/elastic-agent-shipper-client/blob/ce5fac5d4177ece459766ff56d39028ed28da937/api/messages/persisted_index.proto#L16-L20

Reply:
Yes, you're right.

Reply:
Done in https://github.com/elastic/elastic-agent-shipper/pull/76/commits/4d65122671fec178e72a72204fe6a8d0f28d835e